kernel/userspace.c

```diff
@@ -94,6 +98,186 @@ struct perm_ctx {
 	struct k_thread *parent;
 };
 
+#ifdef CONFIG_DYNAMIC_OBJECTS
+struct dyn_obj {
+	struct _k_object kobj;
+	struct rbnode node; /* must be immediately before data member */
+	u8_t data[]; /* The object itself */
+};
+
+struct visit_ctx {
+	_wordlist_cb_func_t func;
+	void *original_context;
+};
+
+extern struct _k_object *_k_object_gperf_find(void *obj);
+extern void _k_object_gperf_wordlist_foreach(_wordlist_cb_func_t func,
+					     void *context);
+
+static int node_lessthan(struct rbnode *a, struct rbnode *b);
+
+static struct rbtree obj_rb_tree = {
+	.lessthan_fn = node_lessthan
+};
+
+/* TODO: incorporate auto-gen with Leandro's patch */
+static size_t obj_size_get(enum k_objects otype)
+{
+	switch (otype) {
+	case K_OBJ_ALERT:
+		return sizeof(struct k_alert);
+	case K_OBJ_MSGQ:
+		return sizeof(struct k_msgq);
+	case K_OBJ_MUTEX:
+		return sizeof(struct k_mutex);
+	case K_OBJ_PIPE:
+		return sizeof(struct k_pipe);
+	case K_OBJ_SEM:
+		return sizeof(struct k_sem);
+	case K_OBJ_STACK:
+		return sizeof(struct k_stack);
+	case K_OBJ_THREAD:
+		return sizeof(struct k_thread);
+	case K_OBJ_TIMER:
+		return sizeof(struct k_timer);
+	default:
+		return sizeof(struct device);
+	}
+}
+
+static int node_lessthan(struct rbnode *a, struct rbnode *b)
+{
+	return a < b;
+}
+
+static inline struct dyn_obj *node_to_dyn_obj(struct rbnode *node)
+{
+	return (struct dyn_obj *)((char *)node -
+				  offsetof(struct dyn_obj, node));
+}
+
+static struct dyn_obj *dyn_object_find(void *obj)
+{
+	struct rbnode *node;
+	struct dyn_obj *ret;
+	int key;
+
+	/* For any dynamically allocated kernel object, the object
+	 * pointer is just a member of the conatining struct dyn_obj,
+	 * so just a little artithmetic is necessary to locate the
+	 * corresponding struct rbnode
+	 */
+	node = (struct rbnode *)((char *)obj - sizeof(struct rbnode));
```

Comment:
Not sure how much of an issue this is, but it's UB to create a pointer to anything that's not a valid object.  Usermode can try to operate, say, on 0xdeadbeef, and this will invoke UB.

Reply:
rb_contains() doesn't deref the node pointer passed in, only its memory address value is considered.
I need to cast this to a struct rbnode to determine whether the node's memory address is in the red/black tree or not.
I suppose I can leave a comment explaining what's going on here? If rb_contains returns true *then* we deref it.

Reply:
Dereferencing a pointer or not does not make a difference: just the act of calculating the pointer to something that does not point to a valid object is UB in C.

Reply:
I'm not sure what to do here.

Reply:
Is that really illegal per C?  The meaning of this code is "there is a struct rbnode stored immediately before the object in memory", and this is a straightforward way of expressing that.

I mean, it would be possible to type this all the way through the stack by defining some k_object_header struct or something that included a struct rbnode at the start and a giant union of all possible k_object types after, and then use CONTAINER_OF.  That would be a whole lot of typing though for not a lot of justification.